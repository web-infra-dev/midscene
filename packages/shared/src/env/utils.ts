import { GlobalConfigManager } from './global-config-manager';
import { ModelConfigManager } from './model-config-manager';
import {
  type GLOBAL_ENV_KEYS,
  MIDSCENE_PREFERRED_LANGUAGE,
  type MODEL_ENV_KEYS,
} from './types';

export const globalModelConfigManager = new ModelConfigManager();

export const globalConfigManager = new GlobalConfigManager();

globalConfigManager.registerModelConfigManager(globalModelConfigManager);
globalModelConfigManager.registerGlobalConfigManager(globalConfigManager);

export const getPreferredLanguage = () => {
  const prefer = globalConfigManager.getEnvConfigValue(
    MIDSCENE_PREFERRED_LANGUAGE,
  );
  if (prefer) {
    return prefer;
  }

  const timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
  const isChina = timeZone === 'Asia/Shanghai';
  return isChina ? 'Chinese' : 'English';
};

export const overrideAIConfig = (
  newConfig: Partial<
    Record<
      (typeof GLOBAL_ENV_KEYS)[number] | (typeof MODEL_ENV_KEYS)[number],
      string
    >
  >,
  extendMode = false, // true: merge with global config, false: override global config
) => {
  globalConfigManager.overrideAIConfig(newConfig, extendMode);
};
