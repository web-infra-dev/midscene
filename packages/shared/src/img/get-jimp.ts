import type Jimp from 'jimp/browser/lib/jimp.js';

const ifInBrowser = typeof window !== 'undefined';
export default async function getJimp(): Promise<typeof Jimp> {
  if (ifInBrowser) {
    await import('jimp/browser/lib/jimp.js');
    return (window as any).Jimp;
  }
  // return Jimp;
  return (await import('jimp')).default;
}
