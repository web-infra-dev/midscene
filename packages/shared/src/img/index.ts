export { imageInfo, imageInfoOfBase64, base64Encoded } from './info';
export {
  trimImage,
  calculateNewDimensions,
  resizeImg,
  transformImgPathToBase64,
  saveBase64Image,
} from './transform';
export { processImageElementInfo } from './box-select';
