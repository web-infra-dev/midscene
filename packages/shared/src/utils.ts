import { sha256 } from 'js-sha256';

export const ifInBrowser = typeof window !== 'undefined';

export function uuid() {
  return Math.random().toString(36).substring(2, 15);
}

const hashMap: Record<string, string> = {}; // id - combined

let frameId = 0;

export function getFrameId(): number {
  return frameId;
}

export function setFrameId(id: number) {
  frameId = id;
}

export function generateHashId(rect: any, content = '') {
  // Combine the input into a string
  const combined = JSON.stringify({
    content,
    rect,
    _midscene_frame_id: getFrameId(),
  });

  // Generates the sha-256 hash value
  let sliceLength = 8;
  let slicedHash = '';
  const hashHex = sha256.create().update(combined).hex();
  while (sliceLength < hashHex.length - 1) {
    slicedHash = hashHex.slice(0, sliceLength);
    if (hashMap[slicedHash] && hashMap[slicedHash] !== combined) {
      sliceLength++;
      continue;
    }
    hashMap[slicedHash] = combined;
    break;
  }
  return slicedHash;
}
