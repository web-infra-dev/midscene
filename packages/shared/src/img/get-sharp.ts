import { ifInNode } from '../utils';

let sharpModule: any = null;

export default async function getSharp(): Promise<any> {
  if (sharpModule) {
    return sharpModule;
  }

  if (!ifInNode) {
    throw new Error('Sharp is only available in Node.js environment');
  }

  try {
    // @ts-ignore
    const sharp = await import('sharp');
    sharpModule = sharp.default;
    return sharpModule;
  } catch (error) {
    throw new Error(
      `Failed to load sharp module: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
