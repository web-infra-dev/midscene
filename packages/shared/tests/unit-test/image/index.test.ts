import { base64Encoded, imageInfo, imageInfoOfBase64 } from '@/img';
import { getFixture } from 'tests/utils';
import { describe, expect, it } from 'vitest';

describe('image utils', () => {
  const image = getFixture('icon.png');
  it('imageInfo', async () => {
    const info = await imageInfo(image);
    expect(info).toMatchSnapshot();
  });

  it('base64Encoded', () => {
    const base64 = base64Encoded(image);
    expect(base64).toMatchSnapshot();

    const headlessBase64 = base64Encoded(image, false);
    expect(headlessBase64).toMatchSnapshot();
  });

  it('base64 + imageInfo', async () => {
    const image = getFixture('icon.png');
    const base64 = base64Encoded(image);
    const info = await imageInfoOfBase64(base64);
    expect(info).toMatchSnapshot();
  });

  it('jpeg + base64 + imageInfo', async () => {
    const image = getFixture('heytea.jpeg');
    const base64 = base64Encoded(image);
    const info = await imageInfoOfBase64(base64);
    expect(info).toMatchSnapshot();
  });

  // it(
  //   'profile',
  //   async () => {
  //     let count = 100;
  //     console.time('alignCoordByTrim');
  //     while (count--) {
  //       const file = getFixture('long-text.png');
  //       await alignCoordByTrim(file, {
  //         left: 440,
  //         top: 50,
  //         width: 200,
  //         height: 150,
  //       });
  //     }
  //     console.timeEnd('alignCoordByTrim');
  //   },
  //   10 * 1000,
  // );
});

// it('align a sub-image', async () => {
//   const file = getFixture('long-text.png');
//   const rect = await alignCoordByTrim(file, {
//     left: 140,
//     top: 50,
//     width: 200,
//     height: 80,
//   });
//   expect(rect).toMatchSnapshot();
// });

// it('align a tiny sub-image', async () => {
//   const file = getFixture('2x2.jpeg');
//   const rect = await alignCoordByTrim(file, {
//     left: 140,
//     top: 50,
//     width: 200,
//     height: 80,
//   });
//   expect(rect).toMatchSnapshot();
// });

// it('align a table style sub-image', async () => {
//   const file = getFixture('table.png');
//   const rect = await alignCoordByTrim(file, {
//     left: 140,
//     top: 50,
//     width: 200,
//     height: 80,
//   });
//   expect(rect).toMatchSnapshot();
// });

// it('illegal center rect, refuse to align', async () => {
//   const file = getFixture('long-text.png'); // 2862x250
//   const rect = await alignCoordByTrim(file, {
//     left: 3000,
//     top: 3000,
//     width: 200,
//     height: 200,
//   });
//   expect(rect).toMatchSnapshot();
// });

// it('align a sub-image with negative coord', async () => {
//   const file = getFixture('long-text.png'); // 2862x250
//   const rect = await alignCoordByTrim(file, {
//     left: -100,
//     top: -100,
//     width: 200,
//     height: 200,
//   });
//   expect(rect).toMatchSnapshot();
// });

// it('align an oversized sub-image', async () => {
//   const file = getFixture('long-text.png'); // 2862x250
//   const rect = await alignCoordByTrim(file, {
//     left: 2860,
//     top: 200,
//     width: 200,
//     height: 200,
//   });
//   expect(rect).toMatchSnapshot();
// });
