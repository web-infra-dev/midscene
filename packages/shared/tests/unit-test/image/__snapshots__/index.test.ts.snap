// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`image utils > align a sub-image 1`] = `
{
  "height": 33,
  "left": 140,
  "top": 97,
  "width": 200,
}
`;

exports[`image utils > align a sub-image with negative coord 1`] = `
{
  "height": 100,
  "left": 0,
  "top": 0,
  "width": 100,
}
`;

exports[`image utils > align a table style sub-image 1`] = `
{
  "height": 57,
  "left": 140,
  "top": 73,
  "width": 200,
}
`;

exports[`image utils > align a tiny sub-image 1`] = `
{
  "height": 80,
  "left": 140,
  "top": 50,
  "width": 200,
}
`;

exports[`image utils > align a tiny sub-image 2`] = `
{
  "height": 80,
  "left": 140,
  "top": 50,
  "width": 200,
}
`;

exports[`image utils > align an oversized sub-image 1`] = `
{
  "height": 50,
  "left": 2860,
  "top": 200,
  "width": 2,
}
`;

exports[`image utils > base64 + imageInfo 1`] = `
{
  "height": 56,
  "width": 68,
}
`;

exports[`image utils > base64Encoded 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEQAAAA4CAYAAABE814IAAAKqGlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUk9kSgO//p4eEltBb6E2QTgApoYcivdoISYBQQggEBTuyqMBaUBHBsiKLAgouKiBrQRSxsAhYwLogi4C6bizYUHk/cAjuvvPeO29y5twv88+dmfufOzkTAMhUlkCQCssCkMbPEob6uNOiY2JpuDGABgRABErAlMXOFDCCgwMAInPr3+X9PQBNr7fNpmP9+/P/KnIcbiYbACgY4XhOJjsN4dOIitkCYRYAqCrErrsySzDN1xCmCpECEX40zYmzLJ7m+BlGo2d8wkM9EFYGAE9isYSJAJD0EDstm52IxCF5ImzB5/D4CCPfgUtaWjoHYSQvMEJ8BAhPx6fHfxcn8W8x4yUxWaxECc+eZUbwnrxMQSor5/98Hf9b0lJFczkMECUlCX1DkRWpCxpISfeXMD9+cdAc8zgz/jOcJPKNmGN2pkfsHHNYnv6SvamLA+Y4gefNlMTJYobPMTfTK2yOhemhklwJQg/GHLOE83lFKRESexKXKYmfmxQeNcfZvMjFc5yZEuY/7+MhsQtFoZL6uXwf9/m83pKzp2V+d14eU7I3KyncV3J21nz9XD5jPmZmtKQ2DtfTa94nQuIvyHKX5BKkBkv8uak+EntmdphkbxZyIef3BkveYTLLL3iOgSfwAgHIhwYigBWwR9QahACvLO6q6TsKPNIFOUJeYlIWjYF0GZfG5LPNF9CsLKxsAJju2dkr8XZgphchRfy8LT8PgEVTCNyatwUi2rgLuT6r5236SF2ySE90/MEWCbNnbdPtBDDIL4EMoAIVoAl0gREwQ2qzA07ADanYDwSBcBADlgM2SAJpQAhWgjVgIygARWAH2APKwSFwBBwDJ0AjaAbnwCVwFdwEPeAueAgGwQh4AcTgPZiEIAgHkSEKpAJpQfqQKWQF0SEXyAsKgEKhGCgOSoT4kAhaA22CiqASqBw6DNVAv0BnoUvQdagXug8NQePQG+gzjIJJMBXWgA3ghTAdZsD+cDi8DE6EM+BcOB/eBpfBlfBxuAm+BN+E78KD8At4AgVQUihFlDbKDEVHeaCCULGoBJQQtQ5ViCpFVaLqUa2oTtRt1CDqJeoTGoumoGloM7QT2hcdgWajM9Dr0MXocvQxdBP6Cvo2eggtRn/DkDHqGFOMI4aJicYkYlZiCjClmGrMGUwH5i5mBPMei8UqYg2x9lhfbAw2GbsaW4w9gG3AtmF7scPYCRwOp4IzxTnjgnAsXBauALcPdxx3EdeHG8F9xEvhtfBWeG98LJ6Pz8OX4mvxF/B9+FH8JEGWoE9wJAQROIQcwnZCFaGVcIswQpgkyhENic7EcGIycSOxjFhP7CA+Ir6VkpLSkXKQCpHiSW2QKpM6KXVNakjqE0meZELyIC0liUjbSEdJbaT7pLdkMtmA7EaOJWeRt5FryJfJT8gfpSnS5tJMaY70eukK6SbpPulXMgQZfRmGzHKZXJlSmVMyt2ReyhJkDWQ9ZFmy62QrZM/K9stOyFHkLOWC5NLkiuVq5a7Ljcnj5A3kveQ58vnyR+Qvyw9TUBRdigeFTdlEqaJ0UEaoWKohlUlNphZRT1C7qWIFeQUbhUiFVQoVCucVBhVRigaKTMVUxe2KjYr3FD8raSgxlLhKW5XqlfqUPiirKbspc5ULlRuU7yp/VqGpeKmkqOxUaVZ5rIpWNVENUV2pelC1Q/WlGlXNSY2tVqjWqPZAHVY3UQ9VX61+RL1LfUJDU8NHQ6CxT+OyxktNRU03zWTN3ZoXNMe1KFouWjyt3VoXtZ7TFGgMWiqtjHaFJtZW1/bVFmkf1u7WntQx1InQydNp0HmsS9Sl6ybo7tZt1xXraekF6q3Rq9N7oE/Qp+sn6e/V79T/YGBoEGWw2aDZYMxQ2ZBpmGtYZ/jIiGzkapRhVGl0xxhrTDdOMT5g3GMCm9iaJJlUmNwyhU3tTHmmB0x7F2AWOCzgL6hc0G9GMmOYZZvVmQ2ZK5oHmOeZN5u/Wqi3MHbhzoWdC79Z2FqkWlRZPLSUt/SzzLNstXxjZWLFtqqwumNNtva2Xm/dYv3axtSGa3PQZsCWYhtou9m23farnb2d0K7ebtxezz7Ofr99P51KD6YX0685YBzcHdY7nHP45GjnmOXY6PiXk5lTilOt09giw0XcRVWLhp11nFnOh50HXWgucS4/uQy6aruyXCtdn7rpunHcqt1GGcaMZMZxxit3C3eh+xn3Dx6OHms92jxRnj6ehZ7dXvJeEV7lXk+8dbwTveu8xT62Pqt92nwxvv6+O337mRpMNrOGKfaz91vrd8Wf5B/mX+7/NMAkQBjQGggH+gXuCny0WH8xf3FzEAhiBu0KehxsGJwR/GsINiQ4pCLkWahl6JrQzjBK2Iqw2rD34e7h28MfRhhFiCLaI2Uil0bWRH6I8owqiRqMXhi9NvpmjGoML6YlFhcbGVsdO7HEa8meJSNLbZcWLL23zHDZqmXXl6suT11+foXMCtaKU3GYuKi42rgvrCBWJWsinhm/P17M9mDvZb/guHF2c8a5ztwS7miCc0JJwliic+KuxPEk16TSpJc8D14573Wyb/Kh5A8pQSlHU6ZSo1Ib0vBpcWln+fL8FP6VdM30Vem9AlNBgWAwwzFjT4ZY6C+szoQyl2W2ZFGR4ahLZCT6QTSU7ZJdkf1xZeTKU6vkVvFXdeWY5GzNGc31zv15NXo1e3X7Gu01G9cMrWWsPbwOWhe/rn297vr89SMbfDYc20jcmLLxtzyLvJK8d5uiNrXma+RvyB/+weeHugLpAmFB/2anzYe2oLfwtnRvtd66b+u3Qk7hjSKLotKiL8Xs4hs/Wv5Y9uPUtoRt3dvtth/cgd3B33Fvp+vOYyVyJbklw7sCdzXtpu0u3P1uz4o910ttSg/tJe4V7R0sCyhr2ae3b8e+L+VJ5Xcr3Csa9qvv37r/wwHOgb6DbgfrD2kcKjr0+SfeTwOHfQ43VRpUlh7BHsk+8qwqsqrzZ/rPNdWq1UXVX4/yjw4eCz12pca+pqZWvXZ7HVwnqhs/vvR4zwnPEy31ZvWHGxQbik6Ck6KTz3+J++Veo39j+yn6qfrT+qf3n6GcKWyCmnKaxM1JzYMtMS29Z/3Otrc6tZ751fzXo+e0z1WcVzi//QLxQv6FqYu5FyfaBG0vLyVeGm5f0f7wcvTlO1dCrnR3+Hdcu+p99XIno/PiNedr5647Xj97g36j+abdzaYu264zv9n+dqbbrrvplv2tlh6HntbeRb0X+lz7Lt32vH31DvPOzbuL7/bei7g30L+0f3CAMzB2P/X+6wfZDyYfbniEeVT4WPZx6RP1J5W/G//eMGg3eH7Ic6jradjTh8Ps4Rd/ZP7xZST/GflZ6ajWaM2Y1di5ce/xnudLno+8ELyYfFnwp9yf+18ZvTr9l9tfXeJo8chr4eupN8VvVd4efWfzrn0ieOLJ+7T3kx8KP6p8PPaJ/qnzc9Tn0cmVX3Bfyr4af2395v/t0VTa1JSAJWTNjAIoROGEBADeHAWAHAMApQcA4pLZmXpGoNn/ATME/hPPzt0zYgfAiTYAgt0A8GybZX3ELOM2awt3A7C1tUTn5t+ZWX1aZI8D4J5jR7cPGNriBP4ps3P8d3X/cwWSqH9b/wVg3QX8UeufjwAAAIplWElmTU0AKgAAAAgABAEaAAUAAAABAAAAPgEbAAUAAAABAAAARgEoAAMAAAABAAIAAIdpAAQAAAABAAAATgAAAAAAAACQAAAAAQAAAJAAAAABAAOShgAHAAAAEgAAAHigAgAEAAAAAQAAAESgAwAEAAAAAQAAADgAAAAAQVNDSUkAAABTY3JlZW5zaG90skJVjwAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAAdRpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDYuMC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6ZXhpZj0iaHR0cDovL25zLmFkb2JlLmNvbS9leGlmLzEuMC8iPgogICAgICAgICA8ZXhpZjpQaXhlbFlEaW1lbnNpb24+NTY8L2V4aWY6UGl4ZWxZRGltZW5zaW9uPgogICAgICAgICA8ZXhpZjpQaXhlbFhEaW1lbnNpb24+Njg8L2V4aWY6UGl4ZWxYRGltZW5zaW9uPgogICAgICAgICA8ZXhpZjpVc2VyQ29tbWVudD5TY3JlZW5zaG90PC9leGlmOlVzZXJDb21tZW50PgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KoAu3GgAAABxpRE9UAAAAAgAAAAAAAAAcAAAAKAAAABwAAAAcAAAA+gvIlt0AAADGSURBVGgF7NIrDsIAFETR1xRBICCQhI9mAVi2gEKxFywLQeGxWLaBJUEhGmj6gQQxCZMaXMWtejc17ckks/nkHTwSSACRxfcA5NcjAAHEBCxZCCAmYMlCADEBSxYCiAlYshBATMCShQBiApYsBBATsGQhgJiAJQtpO0h/NNAn1lUdz0embjrSThrdYU+vyryIPHup/z1at5D1bqt/qIoyTvujuukYL6ax3Kz06n69xeVwVv97AGJigABiApYsBBATsGQhBvIBAAD//8WZ0QUAAAHhSURBVO2UzStEURjGn+tjEml8jQaDaeQrrIYIzWIkSVGKBcrCyn9gw07KxtbeAikpUpSw0VigpIgmJWqGhc9hPgxm1D11X9RczeLQezfnfd7zdu95fj33KEXFlndI9HSN9YvThEOvWJmYF/q7Ir+yEHU9DrF1c+7BzsyG0HoLhYFokTEQLQ8wEAZCCBApdULewmEsj8+RI2ulpdoKe3eTaP67S7VztA+KogiDa1OL8D++CE2LmvZa2OrKRdtzeonduW2h9RbSJaRjpBdJhmTh42DZhYsDt9C0cA53IN2UIdpu1zGO1veF1ltIB6R5sBXZxbnCx2swhM3pVTzfPYmeWtjqK1DTZlfl5+qa3YL37ErT0yOkA2KymdE40KLxEIVyvHmI64jR53sfsgpzUFBlhdVeqpkL+YNYnVzQ9PQK6YBEDTQOOGGy5en1Erk7tuA5/X06oh+UEkhCYiIcQ20wmjNjhnK0vge36yTm+Z8GpQSiHrakoRJljmoYUgxq68v64L3F3tIOHrx3X/Z+05AaiGrIaM76vDdSjWmIpifoD0QuWR+8kd8j+BJQx+Ky/gkgcXEa40sYCAHFQBgIIUAkJ4SBEAJEckIYCCFAJCeEgRACRHJCGAghQCQnhIEQAkR+AOdaTYjLuDCrAAAAAElFTkSuQmCC"`;

exports[`image utils > base64Encoded 2`] = `"iVBORw0KGgoAAAANSUhEUgAAAEQAAAA4CAYAAABE814IAAAKqGlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUk9kSgO//p4eEltBb6E2QTgApoYcivdoISYBQQggEBTuyqMBaUBHBsiKLAgouKiBrQRSxsAhYwLogi4C6bizYUHk/cAjuvvPeO29y5twv88+dmfufOzkTAMhUlkCQCssCkMbPEob6uNOiY2JpuDGABgRABErAlMXOFDCCgwMAInPr3+X9PQBNr7fNpmP9+/P/KnIcbiYbACgY4XhOJjsN4dOIitkCYRYAqCrErrsySzDN1xCmCpECEX40zYmzLJ7m+BlGo2d8wkM9EFYGAE9isYSJAJD0EDstm52IxCF5ImzB5/D4CCPfgUtaWjoHYSQvMEJ8BAhPx6fHfxcn8W8x4yUxWaxECc+eZUbwnrxMQSor5/98Hf9b0lJFczkMECUlCX1DkRWpCxpISfeXMD9+cdAc8zgz/jOcJPKNmGN2pkfsHHNYnv6SvamLA+Y4gefNlMTJYobPMTfTK2yOhemhklwJQg/GHLOE83lFKRESexKXKYmfmxQeNcfZvMjFc5yZEuY/7+MhsQtFoZL6uXwf9/m83pKzp2V+d14eU7I3KyncV3J21nz9XD5jPmZmtKQ2DtfTa94nQuIvyHKX5BKkBkv8uak+EntmdphkbxZyIef3BkveYTLLL3iOgSfwAgHIhwYigBWwR9QahACvLO6q6TsKPNIFOUJeYlIWjYF0GZfG5LPNF9CsLKxsAJju2dkr8XZgphchRfy8LT8PgEVTCNyatwUi2rgLuT6r5236SF2ySE90/MEWCbNnbdPtBDDIL4EMoAIVoAl0gREwQ2qzA07ADanYDwSBcBADlgM2SAJpQAhWgjVgIygARWAH2APKwSFwBBwDJ0AjaAbnwCVwFdwEPeAueAgGwQh4AcTgPZiEIAgHkSEKpAJpQfqQKWQF0SEXyAsKgEKhGCgOSoT4kAhaA22CiqASqBw6DNVAv0BnoUvQdagXug8NQePQG+gzjIJJMBXWgA3ghTAdZsD+cDi8DE6EM+BcOB/eBpfBlfBxuAm+BN+E78KD8At4AgVQUihFlDbKDEVHeaCCULGoBJQQtQ5ViCpFVaLqUa2oTtRt1CDqJeoTGoumoGloM7QT2hcdgWajM9Dr0MXocvQxdBP6Cvo2eggtRn/DkDHqGFOMI4aJicYkYlZiCjClmGrMGUwH5i5mBPMei8UqYg2x9lhfbAw2GbsaW4w9gG3AtmF7scPYCRwOp4IzxTnjgnAsXBauALcPdxx3EdeHG8F9xEvhtfBWeG98LJ6Pz8OX4mvxF/B9+FH8JEGWoE9wJAQROIQcwnZCFaGVcIswQpgkyhENic7EcGIycSOxjFhP7CA+Ir6VkpLSkXKQCpHiSW2QKpM6KXVNakjqE0meZELyIC0liUjbSEdJbaT7pLdkMtmA7EaOJWeRt5FryJfJT8gfpSnS5tJMaY70eukK6SbpPulXMgQZfRmGzHKZXJlSmVMyt2ReyhJkDWQ9ZFmy62QrZM/K9stOyFHkLOWC5NLkiuVq5a7Ljcnj5A3kveQ58vnyR+Qvyw9TUBRdigeFTdlEqaJ0UEaoWKohlUlNphZRT1C7qWIFeQUbhUiFVQoVCucVBhVRigaKTMVUxe2KjYr3FD8raSgxlLhKW5XqlfqUPiirKbspc5ULlRuU7yp/VqGpeKmkqOxUaVZ5rIpWNVENUV2pelC1Q/WlGlXNSY2tVqjWqPZAHVY3UQ9VX61+RL1LfUJDU8NHQ6CxT+OyxktNRU03zWTN3ZoXNMe1KFouWjyt3VoXtZ7TFGgMWiqtjHaFJtZW1/bVFmkf1u7WntQx1InQydNp0HmsS9Sl6ybo7tZt1xXraekF6q3Rq9N7oE/Qp+sn6e/V79T/YGBoEGWw2aDZYMxQ2ZBpmGtYZ/jIiGzkapRhVGl0xxhrTDdOMT5g3GMCm9iaJJlUmNwyhU3tTHmmB0x7F2AWOCzgL6hc0G9GMmOYZZvVmQ2ZK5oHmOeZN5u/Wqi3MHbhzoWdC79Z2FqkWlRZPLSUt/SzzLNstXxjZWLFtqqwumNNtva2Xm/dYv3axtSGa3PQZsCWYhtou9m23farnb2d0K7ebtxezz7Ofr99P51KD6YX0685YBzcHdY7nHP45GjnmOXY6PiXk5lTilOt09giw0XcRVWLhp11nFnOh50HXWgucS4/uQy6aruyXCtdn7rpunHcqt1GGcaMZMZxxit3C3eh+xn3Dx6OHms92jxRnj6ehZ7dXvJeEV7lXk+8dbwTveu8xT62Pqt92nwxvv6+O337mRpMNrOGKfaz91vrd8Wf5B/mX+7/NMAkQBjQGggH+gXuCny0WH8xf3FzEAhiBu0KehxsGJwR/GsINiQ4pCLkWahl6JrQzjBK2Iqw2rD34e7h28MfRhhFiCLaI2Uil0bWRH6I8owqiRqMXhi9NvpmjGoML6YlFhcbGVsdO7HEa8meJSNLbZcWLL23zHDZqmXXl6suT11+foXMCtaKU3GYuKi42rgvrCBWJWsinhm/P17M9mDvZb/guHF2c8a5ztwS7miCc0JJwliic+KuxPEk16TSpJc8D14573Wyb/Kh5A8pQSlHU6ZSo1Ib0vBpcWln+fL8FP6VdM30Vem9AlNBgWAwwzFjT4ZY6C+szoQyl2W2ZFGR4ahLZCT6QTSU7ZJdkf1xZeTKU6vkVvFXdeWY5GzNGc31zv15NXo1e3X7Gu01G9cMrWWsPbwOWhe/rn297vr89SMbfDYc20jcmLLxtzyLvJK8d5uiNrXma+RvyB/+weeHugLpAmFB/2anzYe2oLfwtnRvtd66b+u3Qk7hjSKLotKiL8Xs4hs/Wv5Y9uPUtoRt3dvtth/cgd3B33Fvp+vOYyVyJbklw7sCdzXtpu0u3P1uz4o910ttSg/tJe4V7R0sCyhr2ae3b8e+L+VJ5Xcr3Csa9qvv37r/wwHOgb6DbgfrD2kcKjr0+SfeTwOHfQ43VRpUlh7BHsk+8qwqsqrzZ/rPNdWq1UXVX4/yjw4eCz12pca+pqZWvXZ7HVwnqhs/vvR4zwnPEy31ZvWHGxQbik6Ck6KTz3+J++Veo39j+yn6qfrT+qf3n6GcKWyCmnKaxM1JzYMtMS29Z/3Otrc6tZ751fzXo+e0z1WcVzi//QLxQv6FqYu5FyfaBG0vLyVeGm5f0f7wcvTlO1dCrnR3+Hdcu+p99XIno/PiNedr5647Xj97g36j+abdzaYu264zv9n+dqbbrrvplv2tlh6HntbeRb0X+lz7Lt32vH31DvPOzbuL7/bei7g30L+0f3CAMzB2P/X+6wfZDyYfbniEeVT4WPZx6RP1J5W/G//eMGg3eH7Ic6jradjTh8Ps4Rd/ZP7xZST/GflZ6ajWaM2Y1di5ce/xnudLno+8ELyYfFnwp9yf+18ZvTr9l9tfXeJo8chr4eupN8VvVd4efWfzrn0ieOLJ+7T3kx8KP6p8PPaJ/qnzc9Tn0cmVX3Bfyr4af2395v/t0VTa1JSAJWTNjAIoROGEBADeHAWAHAMApQcA4pLZmXpGoNn/ATME/hPPzt0zYgfAiTYAgt0A8GybZX3ELOM2awt3A7C1tUTn5t+ZWX1aZI8D4J5jR7cPGNriBP4ps3P8d3X/cwWSqH9b/wVg3QX8UeufjwAAAIplWElmTU0AKgAAAAgABAEaAAUAAAABAAAAPgEbAAUAAAABAAAARgEoAAMAAAABAAIAAIdpAAQAAAABAAAATgAAAAAAAACQAAAAAQAAAJAAAAABAAOShgAHAAAAEgAAAHigAgAEAAAAAQAAAESgAwAEAAAAAQAAADgAAAAAQVNDSUkAAABTY3JlZW5zaG90skJVjwAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAAdRpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDYuMC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6ZXhpZj0iaHR0cDovL25zLmFkb2JlLmNvbS9leGlmLzEuMC8iPgogICAgICAgICA8ZXhpZjpQaXhlbFlEaW1lbnNpb24+NTY8L2V4aWY6UGl4ZWxZRGltZW5zaW9uPgogICAgICAgICA8ZXhpZjpQaXhlbFhEaW1lbnNpb24+Njg8L2V4aWY6UGl4ZWxYRGltZW5zaW9uPgogICAgICAgICA8ZXhpZjpVc2VyQ29tbWVudD5TY3JlZW5zaG90PC9leGlmOlVzZXJDb21tZW50PgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KoAu3GgAAABxpRE9UAAAAAgAAAAAAAAAcAAAAKAAAABwAAAAcAAAA+gvIlt0AAADGSURBVGgF7NIrDsIAFETR1xRBICCQhI9mAVi2gEKxFywLQeGxWLaBJUEhGmj6gQQxCZMaXMWtejc17ckks/nkHTwSSACRxfcA5NcjAAHEBCxZCCAmYMlCADEBSxYCiAlYshBATMCShQBiApYsBBATsGQhgJiAJQtpO0h/NNAn1lUdz0embjrSThrdYU+vyryIPHup/z1at5D1bqt/qIoyTvujuukYL6ax3Kz06n69xeVwVv97AGJigABiApYsBBATsGQhBvIBAAD//8WZ0QUAAAHhSURBVO2UzStEURjGn+tjEml8jQaDaeQrrIYIzWIkSVGKBcrCyn9gw07KxtbeAikpUpSw0VigpIgmJWqGhc9hPgxm1D11X9RczeLQezfnfd7zdu95fj33KEXFlndI9HSN9YvThEOvWJmYF/q7Ir+yEHU9DrF1c+7BzsyG0HoLhYFokTEQLQ8wEAZCCBApdULewmEsj8+RI2ulpdoKe3eTaP67S7VztA+KogiDa1OL8D++CE2LmvZa2OrKRdtzeonduW2h9RbSJaRjpBdJhmTh42DZhYsDt9C0cA53IN2UIdpu1zGO1veF1ltIB6R5sBXZxbnCx2swhM3pVTzfPYmeWtjqK1DTZlfl5+qa3YL37ErT0yOkA2KymdE40KLxEIVyvHmI64jR53sfsgpzUFBlhdVeqpkL+YNYnVzQ9PQK6YBEDTQOOGGy5en1Erk7tuA5/X06oh+UEkhCYiIcQ20wmjNjhnK0vge36yTm+Z8GpQSiHrakoRJljmoYUgxq68v64L3F3tIOHrx3X/Z+05AaiGrIaM76vDdSjWmIpifoD0QuWR+8kd8j+BJQx+Ky/gkgcXEa40sYCAHFQBgIIUAkJ4SBEAJEckIYCCFAJCeEgRACRHJCGAghQCQnhIEQAkR+AOdaTYjLuDCrAAAAAElFTkSuQmCC"`;

exports[`image utils > illegal center rect, refuse to align 1`] = `
{
  "height": -1,
  "left": 0,
  "top": 0,
  "width": -1,
}
`;

exports[`image utils > imageInfo 1`] = `
{
  "height": 56,
  "width": 68,
}
`;

exports[`image utils > jpeg + base64 + imageInfo 1`] = `
{
  "height": 905,
  "width": 400,
}
`;

exports[`image utils > trim image 1`] = `
{
  "height": 70,
  "trimOffsetLeft": -93,
  "trimOffsetTop": -97,
  "width": 1151,
}
`;

exports[`image utils > trim image 2`] = `
{
  "height": 555,
  "trimOffsetLeft": 0,
  "trimOffsetTop": -21,
  "width": 1486,
}
`;

exports[`image utils > trim image 3`] = `
{
  "height": 296,
  "trimOffsetLeft": 0,
  "trimOffsetTop": 0,
  "width": 482,
}
`;
